File: src/main/java/com/freya02/botcommands/api/pagination/paginator/BasicPaginatorBuilder.java
package com.freya02.botcommands.api.pagination.paginator;

import com.freya02.botcommands.api.components.Components;
import com.freya02.botcommands.api.pagination.BasicPagination;
import com.freya02.botcommands.api.pagination.BasicPaginationBuilder;
import com.freya02.botcommands.api.pagination.PaginatorSupplier;
import com.freya02.botcommands.api.utils.ButtonContent;
import org.jetbrains.annotations.NotNull;

/**
 * Provides base for a paginator builder
 *
 * @param <T> Type of the implementor
 * @param <R> Type of the implementor {@link #build()} return type
 */
@SuppressWarnings("unchecked")
public abstract class BasicPaginatorBuilder<T extends BasicPaginationBuilder<T, R>, R extends BasicPagination<R>> extends BasicPaginationBuilder<T, R> {
	private static final ButtonContent DEFAULT_FIRST_CONTENT = ButtonContent.withShortcode("rewind");
	private static final ButtonContent DEFAULT_PREVIOUS_CONTENT = ButtonContent.withShortcode("arrow_backward");
	private static final ButtonContent DEFAULT_NEXT_CONTENT = ButtonContent.withShortcode("arrow_forward");
	private static final ButtonContent DEFAULT_LAST_CONTENT = ButtonContent.withShortcode("fast_forward");
	private static final ButtonContent DEFAULT_DELETE_CONTENT = ButtonContent.withShortcode("wastebasket");

	protected PaginatorSupplier<R> paginatorSupplier;

	protected ButtonContent firstContent = DEFAULT_FIRST_CONTENT;
	protected ButtonContent previousContent = DEFAULT_PREVIOUS_CONTENT;
	protected ButtonContent nextContent = DEFAULT_NEXT_CONTENT;
	protected ButtonContent lastContent = DEFAULT_LAST_CONTENT;
	protected ButtonContent deleteContent = DEFAULT_DELETE_CONTENT;

	protected boolean hasDeleteButton;

	public BasicPaginatorBuilder(@NotNull Components componentsService) {
		super(componentsService);
	}

	/**
	 * Sets the {@link PaginatorSupplier} for this paginator
	 * <br>This is what supplies the pages dynamically for this paginator
	 *
	 * @param paginatorSupplier The {@link PaginatorSupplier} for this paginator
	 * @return This builder for chaining convenience
	 */
	public T setPaginatorSupplier(@NotNull PaginatorSupplier<R> paginatorSupplier) {
		this.paginatorSupplier = paginatorSupplier;

		return (T) this;
	}

	/**
	 * Specifies whether this paginator should have a delete button
	 * <br>Note that this button <b><i>does</i></b> cleanup used components
	 *
	 * @param hasDeleteButton <code>true</code> if the delete button has to appear
	 * @return This builder for chaining convenience
	 */
	public T useDeleteButton(boolean hasDeleteButton) {
		this.hasDeleteButton = hasDeleteButton;

		return (T) this;
	}

	/**
	 * Sets the content for the button which goes to the first page
	 *
	 * @param firstContent The {@link ButtonContent} for this button
	 * @return This builder for chaining convenience
	 */
	public T setFirstContent(ButtonContent firstContent) {
		this.firstContent = firstContent;

		return (T) this;
	}

	/**
	 * Sets the content for the button which goes to the previous page
	 *
	 * @param previousContent The {@link ButtonContent} for this button
	 * @return This builder for chaining convenience
	 */
	public T setPreviousContent(ButtonContent previousContent) {
		this.previousContent = previousContent;

		return (T) this;
	}

	/**
	 * Sets the content for the button which goes to the next page
	 *
	 * @param nextContent The {@link ButtonContent} for this button
	 * @return This builder for chaining convenience
	 */
	public T setNextContent(ButtonContent nextContent) {
		this.nextContent = nextContent;

		return (T) this;
	}

	/**
	 * Sets the content for the button which goes to the last page
	 *
	 * @param lastContent The {@link ButtonContent} for this button
	 * @return This builder for chaining convenience
	 */
	public T setLastContent(ButtonContent lastContent) {
		this.lastContent = lastContent;

		return (T) this;
	}

	/**
	 * Sets the content for the button which deletes this paginator
	 *
	 * @param deleteContent The {@link ButtonContent} for this button
	 * @return This builder for chaining convenience
	 */
	public T setDeleteContent(ButtonContent deleteContent) {
		this.deleteContent = deleteContent;

		return (T) this;
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/pagination/paginator/BasicPaginatorBuilder.kt
package io.github.freya022.botcommands.api.pagination.paginator

import io.github.freya022.botcommands.api.components.Components
import io.github.freya022.botcommands.api.pagination.BasicPagination
import io.github.freya022.botcommands.api.pagination.BasicPaginationBuilder
import io.github.freya022.botcommands.api.utils.ButtonContent

/**
 * Most basic paginator builder.
 *
 * @param T Type of the pagination builder
 * @param R Type of the built pagination
 */
abstract class BasicPaginatorBuilder<T : BasicPaginationBuilder<T, R>, R : BasicPagination<R>>(
    componentsService: Components
) : BasicPaginationBuilder<T, R>(componentsService) {
    var firstContent: ButtonContent = DEFAULT_FIRST_CONTENT
        private set
    var previousContent: ButtonContent = DEFAULT_PREVIOUS_CONTENT
        private set
    var nextContent: ButtonContent = DEFAULT_NEXT_CONTENT
        private set
    var lastContent: ButtonContent = DEFAULT_LAST_CONTENT
        private set
    var deleteContent: ButtonContent = DEFAULT_DELETE_CONTENT
        private set

    var hasDeleteButton: Boolean = false
        private set

    /**
     * Specifies whether this paginator should have a delete button.
     *
     * The button can only be used by those satisfying [the constraints][setConstraints].
     *
     * @param hasDeleteButton `true` if the delete button has to appear
     *
     * @return This builder for chaining convenience
     *
     * @see setConstraints
     */
    fun useDeleteButton(hasDeleteButton: Boolean): T = config {
        this.hasDeleteButton = hasDeleteButton
    }

    /**
     * Sets the content for the button which goes to the first page
     *
     * @param firstContent The [ButtonContent] for this button
     *
     * @return This builder for chaining convenience
     */
    fun setFirstContent(firstContent: ButtonContent): T = config {
        this.firstContent = firstContent
    }

    /**
     * Sets the content for the button which goes to the previous page
     *
     * @param previousContent The [ButtonContent] for this button
     *
     * @return This builder for chaining convenience
     */
    fun setPreviousContent(previousContent: ButtonContent): T = config {
        this.previousContent = previousContent
    }

    /**
     * Sets the content for the button which goes to the next page
     *
     * @param nextContent The [ButtonContent] for this button
     *
     * @return This builder for chaining convenience
     */
    fun setNextContent(nextContent: ButtonContent): T = config {
        this.nextContent = nextContent
    }

    /**
     * Sets the content for the button which goes to the last page
     *
     * @param lastContent The [ButtonContent] for this button
     *
     * @return This builder for chaining convenience
     */
    fun setLastContent(lastContent: ButtonContent): T = config {
        this.lastContent = lastContent
    }

    /**
     * Sets the content for the button which deletes this paginator
     *
     * @param deleteContent The [ButtonContent] for this button
     *
     * @return This builder for chaining convenience
     */
    fun setDeleteContent(deleteContent: ButtonContent): T = config {
        this.deleteContent = deleteContent
    }

    companion object {
        // TODO put in PaginatorBuilder
        @JvmStatic
        var DEFAULT_FIRST_CONTENT: ButtonContent = ButtonContent.withShortcode("rewind")
        @JvmStatic
        var DEFAULT_PREVIOUS_CONTENT: ButtonContent = ButtonContent.withShortcode("arrow_backward")
        @JvmStatic
        var DEFAULT_NEXT_CONTENT: ButtonContent = ButtonContent.withShortcode("arrow_forward")
        @JvmStatic
        var DEFAULT_LAST_CONTENT: ButtonContent = ButtonContent.withShortcode("fast_forward")
        @JvmStatic
        var DEFAULT_DELETE_CONTENT: ButtonContent = ButtonContent.withShortcode("wastebasket")
    }
}
